File: LeetCode Questions/206. Reverse Linked List/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head):
        # Iterative Solution
        previous, current = None, head
        while current:
            next = current.next
            current.next = previous
            previous = current
            current = next
            
        return previous

        # Recursive solution to reverse a linked list
        if head is None or head.next is None:
            return head
        reversed_head = self.reverseList(head.next)
        head.next.next = head
        head.next = None
        return reversed_head

File: LeetCode Questions/206. Reverse Linked List/solution.js

/**
 * @param {ListNode} head
 * @return {ListNode}
 */

// Iterative solution - O(n) time, O(1) space
var reverseList = function(head) {
    let previous = None
    let current = head
        
    while(current != null) {
        next = current.next
        current.next = previous
        previous = current
        current = next
    }

    return previous
};

// Recursive solution - O(n) time, O(n) space
var reverseList = function(head) {
    if(head === null || head.next === null) {
        return head
    }

    var reversedList = reverseList(head.next)
    head.next.next = head
    head.next = null
    return reversedList
};